README.md
FixedPoint
============

For modeling fixed point signed and unsigned data types and having nice function for printing hex and binary forms.

Install
-------

    gem install fixed_point

Usage
-----

Checkout the examples folder, but here are a few:
 
    require 'fixed_point'

    #Create fixed point format, Signed, 12 integer bits, 4 fractional bits
    format  = FixedPoint::Format.new(1, 12, 4) 

    #Create fixed_point with value 1024.75
    fix_num = FixedPoint::Number.new(1024.75, format )

    puts fix_num.to_f # Float
    puts fix_num.to_h # Hexadecimal
    puts fix_num.to_b # Binary
    
Building tables:

    require 'fixed_point'
    
    #Signed 8 bit, (4 bit Integer, 4 bit fractional)
    format = FixedPoint::Format.new(1,4,4)

    table_for_conversion = [ 
      2.7505, 2.5, 2.25, 2, 0.5, 0, -0.5, -4.5   
    ]

    table_for_conversion.each do |num|
      fixt   = FixedPoint::Number.new(num, format, '_')
      puts "#{format.width}'b#{fixt.to_b}  // Quantised %7.4f  Source %7.4f" % [ fixt.to_f, fixt.source]
    end

Returns:

    8'b0010_1100  // Quantised  2.7500  Source  2.7505
    8'b0010_1000  // Quantised  2.5000  Source  2.5000
    8'b0010_0100  // Quantised  2.2500  Source  2.2500
    8'b0010_0000  // Quantised  2.0000  Source  2.0000
    8'b0000_1000  // Quantised  0.5000  Source  0.5000
    8'b0000_0000  // Quantised  0.0000  Source  0.0000
    8'b1111_1000  // Quantised -0.5000  Source -0.5000
    8'b1011_1000  // Quantised -4.5000  Source -4.5000




TODO
----


LICENSE
-------

See the LICENSE file


HISTORY.md
HISTORY/Changelog
=================

0.2.1
-----

 Added hex= method for setting fixed point data via hex

0.2.0
-----

 Added to_s method on numbers. Output is now the same as calling .to_f.to_s

0.1.1
-----

 Depreciated .fraction method, use .frac instead

0.1.0
-----

 Improved API, to_b, to_h, to_i, to_f
 Working hex output

0.0.1
-----

 Define number of integer and fractional bits with real/integer value input. 



spec/spec_helper.rb
$LOAD_PATH.unshift File.join(File.dirname(__FILE__), '..', 'lib')

require 'rspec'
require 'fixed_point'


spec/fixed_point_spec.rb
require 'spec_helper'

# def initialize(number, Format.new(signed=1, int_bits=12, frac_bits=20), decimal_mark=".")
# source Input Number 
# to_f   quantised value as float
# to_i   quantised value as integer
# frac   fractional part of quantised value
# to_h   quantised value as string formatted as hex
# to_b   quantised value as string formatted as binary
#

describe FixedPoint do

  it "Returns 0.0 for initalise of 0" do
    fixt = FixedPoint::Number.new(0)

    fixt.source.should    == 0.0
    fixt.to_f.should      == 0.0
    fixt.to_s.should      == "0.0"
    fixt.to_i.should      == 0
    fixt.frac.should      == 0.0
    fixt.to_h.should      == "00000000"
    fixt.to_b.should      == "000000000000.00000000000000000000"  
  end

  it "Integers 0 " do
    format = FixedPoint::Format.new(1,8,0)
    fixt   = FixedPoint::Number.new(7.0, format, "_")

    fixt.source.should   == 7.0
    fixt.to_f.should     == 7.0
    fixt.to_s.should     == "7.0"
    fixt.to_i.should     == 7
    fixt.frac.should     == 0.0
    fixt.to_h.should     == "07"
    fixt.to_b.should     == "00000111"
  end

  it "Different Decimal Mark _ instead of ." do
    format = FixedPoint::Format.new(1,12,20)
    fixt   = FixedPoint::Number.new(0, format, "_")

    fixt.source.should   == 0.0
    fixt.to_f.should     == 0.0
    fixt.to_s.should     == "0.0"
    fixt.to_i.should     == 0
    fixt.frac.should     == 0.0
    fixt.to_h.should     == "00000000"
    fixt.to_b.should     == "000000000000_00000000000000000000"
  end

  (1...20).to_a.reverse_each do |x|
    int_bits = 12
    it "Returns 0.0 for initalise of 0 with #{x} fractional bits" do
      format = FixedPoint::Format.new(1, int_bits, x)
      fixt   = FixedPoint::Number.new(0, format)
      #fixt = FixedPoint::Number.new(0,1,int_bits,x)

      fixt.source.should   == 0.0
      fixt.to_f.should     == 0.0
      fixt.to_s.should     == "0.0"
      fixt.to_i.should     == 0
      fixt.frac.should     == 0.0
      #Calculate hex length and fill with 0's
      hex = ""
      hex_length = ((x.to_f+int_bits.to_f)/4).ceil
      hex_length.times { hex += "0" }
      fixt.to_h.should     == hex
      #Calculate binary fractional length and fill with 0's
      lsbs = ""
      x.times { lsbs += "0" }
      fixt.to_b.should     == "000000000000.#{lsbs}"
    end
  end


  it "Zero fractional bits " do
    format = FixedPoint::Format.new(1,12,0)
    fixt   = FixedPoint::Number.new(0, format, "_")

    fixt.source.should   == 0.0
    fixt.to_f.should     == 0.0
    fixt.to_s.should     == "0.0"
    fixt.to_i.should     == 0
    fixt.frac.should     == 0.0
    fixt.to_h.should     == "000"
    fixt.to_b.should     == "000000000000"
  end


  it "returns 2.5 for initalise of 2.5" do
    fixt = FixedPoint::Number.new(2.5)

    fixt.source.should   == 2.5
    fixt.to_f.should     == 2.5
    fixt.to_s.should     == "2.5"
    fixt.to_i.should     == 2
    fixt.frac.should     == 0.5
    fixt.to_h.should     == "00280000"
    fixt.to_b.should     == "000000000010.10000000000000000000"
  end


  it "Truncates fractional numbers correctly" do
    format = FixedPoint::Format.new(1,12,1)
    fixt   = FixedPoint::Number.new(2.501, format)

    fixt.source.should   == 2.501
    fixt.to_f.should     == 2.5
    fixt.to_s.should     == "2.5"
    fixt.to_i.should     == 2
    fixt.frac.should     == 0.5
    fixt.to_h.should     == "0005"
    fixt.to_b.should     == "000000000010.1"
  end

  ##############################################
  ###   Overflow Section
  ##############################################
  it "Forced Overflow 4 Int Bits, 0 fractional bits " do
    format = FixedPoint::Format.new(1, 4, 0)
    fixt   = FixedPoint::Number.new(2**3, format, "_")

    fixt.source.should     == 2**3
    fixt.to_f.should       == 2**3-1
    fixt.to_s.should       == "7.0"
    fixt.to_i.should       == 2**3-1
    fixt.frac.should       == 0.0
    fixt.to_h.should       == "7"
    fixt.to_b.should       == "0111"
    fixt.overflow?.should  == true
    fixt.underflow?.should == false
  end

  it "Forced Overflow Zero fractional bits " do
    format = FixedPoint::Format.new(1, 12, 0)
    fixt   = FixedPoint::Number.new(2**11, format, "_")

    fixt.source.should    == 2**11
    fixt.to_f.should      == 2**11-1
    fixt.to_s.should      == "2047.0"
    fixt.to_i.should      == 2**11-1
    fixt.frac.should      == 0.0
    fixt.to_h.should      == "7ff"
    fixt.to_b.should      == "011111111111"
    fixt.overflow?.should  == true
    fixt.underflow?.should == false
  end

  it "Forced Overflow 4 fractional bits " do
    format = FixedPoint::Format.new(1, 12, 4) 
    fixt   = FixedPoint::Number.new(2**11, format, ".")

    max_fractional_value  = (1.0/2) + (1.0/4) + (1.0/8) + (1.0/16)

    fixt.source.should    == 2**11
    fixt.to_f.should      == 2**11 -1 + max_fractional_value
    fixt.to_s.should      == "2047.9375"
    fixt.to_i.should      == 2**11 -1
    fixt.frac.should      == max_fractional_value
    fixt.to_h.should      == "7fff"
    fixt.to_b.should      == "011111111111.1111"
    fixt.overflow?.should  == true
    fixt.underflow?.should == false
  end

  it "Forced Overflow  Large overflow" do
    format = FixedPoint::Format.new(1, 12, 4) 
    fixt   = FixedPoint::Number.new(2**17, format, ".")

    fixt.source.should    == 2**17

    max_fractional_value = (1.0/2) + (1.0/4) + (1.0/8) + (1.0/16)

    fixt.to_f.should      == 2**11 -1 + max_fractional_value
    fixt.to_s.should      == "2047.9375"
    fixt.to_i.should      == 2**11 -1
    fixt.frac.should      == max_fractional_value
    fixt.to_h.should      == "7fff"
    fixt.to_b.should      == "011111111111.1111"
    fixt.overflow?.should  == true
    fixt.underflow?.should == false
  end

  ##############################################
  ###   Underflow Section
  ##############################################
  #Max Negative value
  it "Forced Overflow Zero fractional bits " do
    format = FixedPoint::Format.new(1, 12, 0) 
    fixt   = FixedPoint::Number.new(-2**11, format, "_")

    fixt.source.should    == -2**11
    fixt.to_f.should      == -2**11
    fixt.to_s.should      == "-2048"
    fixt.to_i.should      == -2**11
    fixt.frac.should      == 0.0
    fixt.to_h.should      == "800"
    fixt.to_b.should      == "100000000000"
    fixt.overflow?.should  == false
    fixt.underflow?.should == false
  end

  it "Forced Overflow Zero fractional bits " do
    format = FixedPoint::Format.new(1, 12, 0) 
    fixt   = FixedPoint::Number.new(-2**11-1, format,"_")

    fixt.source.should    == -2**11-1
    fixt.to_f.should      == -2**11
    fixt.to_s.should      == "-2048"
    fixt.to_i.should      == -2**11
    fixt.frac.should      == 0.0
    fixt.to_h.should      == "800"
    fixt.to_b.should      == "100000000000"
    fixt.overflow?.should  == false
    fixt.underflow?.should == true
  end

  it "Forced Overflow 4 fractional bits " do
    format = FixedPoint::Format.new(1, 12, 4) 
    fixt   = FixedPoint::Number.new(-2**11-1, format, ".")

    fixt.source.should    == -2**11-1
    fixt.to_f.should      == -2**11
    fixt.to_s.should      == "-2048"
    fixt.to_i.should      == -2**11
    fixt.frac.should      == 0.0
    fixt.to_h.should      == "8000"
    fixt.to_b.should      == "100000000000.0000"
    fixt.overflow?.should  == false
    fixt.underflow?.should == true
  end

  it "Forced Overflow  Large overflow" do
    format = FixedPoint::Format.new(1, 12, 4) 
    fixt   = FixedPoint::Number.new(-2**17, format, ".")


    fixt.source.should    == -2**17
    fixt.to_f.should      == -2**11
    fixt.to_s.should      == "-2048"
    fixt.to_i.should      == -2**11
    fixt.frac.should      == 0.0
    fixt.to_h.should      == "8000"
    fixt.to_b.should      == "100000000000.0000"
    fixt.overflow?.should  == false
    fixt.underflow?.should == true
  end


  it "Creating via binary form 011_01" do
    fixt = FixedPoint::Number.new(0)

    fixt.binary = "011_01"
    fixt.source.should    == 3.25
    fixt.to_f.should      == 3.25
    fixt.to_s.should      == "3.25"
    fixt.to_i.should      == 3
    fixt.frac.should      == 0.25
    fixt.to_h.should      == "0d"
    fixt.to_b.should      == "011_01"
    fixt.overflow?.should  == false
    fixt.underflow?.should == false
  end

  it "Creating via binary form 0_1" do
    fixt = FixedPoint::Number.new(0)

    fixt.binary = "0_1"
    fixt.source.should    == 0.5
    fixt.to_f.should      == 0.5
    fixt.to_s.should      == "0.5"
    fixt.to_i.should      == 0
    fixt.frac.should      == 0.5
    fixt.to_h.should      == "1"
    fixt.to_b.should      == "0_1"
    fixt.overflow?.should  == false
    fixt.underflow?.should == false
  end

  it "Creating Signed via binary form 1000_1" do
    fixt = FixedPoint::Number.new(0)

    fixt.binary = "1000_1"
    fixt.source.should    == -7.5
    fixt.to_f.should      == -7.5
    fixt.to_s.should      == "-7.5"
    fixt.to_i.should      == -7
    fixt.frac.should      == -0.5
    fixt.to_h.should      == "11"
    fixt.to_b.should      == "1000_1"
    fixt.overflow?.should  == false
    fixt.underflow?.should == false
  end

  it "Creating Unsigned via binary form 1000_1" do
    format = FixedPoint::Format.new(0, 12, 20) 
    fixt = FixedPoint::Number.new(0,format)

    fixt.binary = "1000_1"
    fixt.source.should    == 8.5
    fixt.to_f.should      == 8.5
    fixt.to_s.should      == "8.5"
    fixt.to_i.should      == 8
    fixt.frac.should      == 0.5
    fixt.to_h.should      == "11"
    fixt.to_b.should      == "1000_1"
    fixt.overflow?.should  == false
    fixt.underflow?.should == false
  end

  it "Creating Unsigned binary form 1000_1 using Fixdt datatype" do
    format = FixedPoint::Fixdt.new(0, 32, 20)  #Signed,width,frac_bits
    fixt   = FixedPoint::Number.new(0,format)

    fixt.binary = "1000_1"
    fixt.source.should    == 8.5
    fixt.to_f.should      == 8.5
    fixt.to_s.should      == "8.5"
    fixt.to_i.should      == 8
    fixt.frac.should      == 0.5
    fixt.to_h.should      == "11"
    fixt.to_b.should      == "1000_1"
    fixt.overflow?.should  == false
    fixt.underflow?.should == false
  end

  #need overflow test
  #need underflow test
  #Integer only binary test
end


spec/fixed_point_hex_spec.rb
require 'spec_helper'

describe FixedPoint do
  it "Create 4 bit hex 7" do
    format = FixedPoint::Fixdt.new(0, 4, 0)
    fixt = FixedPoint::Number.new(0, format)
    
    fixt.hex = "7"
    fixt.source.should    == 7.0
    fixt.to_f.should      == 7.0
    fixt.to_s.should      == "7.0"
    fixt.to_i.should      == 7
    fixt.frac.should      == 0.0
    fixt.to_h.should      == "7"
    fixt.to_b.should      == "0111"
    fixt.overflow?.should  == false
    fixt.underflow?.should == false
  end
  
  it "Create 4 bit hex F" do
    format = FixedPoint::Fixdt.new(0, 4, 0)
    fixt = FixedPoint::Number.new(0, format)
    
    fixt.hex = "F"
    fixt.source.should    == 15.0
    fixt.to_f.should      == 15.0
    fixt.to_s.should      == "15.0"
    fixt.to_i.should      == 15
    fixt.frac.should      == 0.0
    fixt.to_h.should      == "f"
    fixt.to_b.should      == "1111"
    fixt.overflow?.should  == false
    fixt.underflow?.should == false
  end
  
  it "Create 4 bit hex F Signed" do
    format = FixedPoint::Fixdt.new(1, 4, 0)
    fixt = FixedPoint::Number.new(0, format)
    
    fixt.hex = "F"
    fixt.source.should    == -1.0
    fixt.to_f.should      == -1.0
    fixt.to_s.should      == "-1.0"
    fixt.to_i.should      == -1
    fixt.frac.should      == 0.0
    fixt.to_h.should      == "f"
    fixt.to_b.should      == "1111"
    fixt.overflow?.should  == false
    fixt.underflow?.should == false
  end
  
  it "Create 4 bit hex F Signed" do
    format = FixedPoint::Fixdt.new(1, 4, 0)
    fixt = FixedPoint::Number.new(0, format)
    
    fixt.hex = "F"
    fixt.source.should    == -1.0
    fixt.to_f.should      == -1.0
    fixt.to_s.should      == "-1.0"
    fixt.to_i.should      == -1
    fixt.frac.should      == 0.0
    fixt.to_h.should      == "f"
    fixt.to_b.should      == "1111"
    fixt.overflow?.should  == false
    fixt.underflow?.should == false
  end

  it "Create 8 bit Signed hex f0 [1,4,4] = -1.0" do
    format = FixedPoint::Format.new(1, 4, 4)
    fixt   = FixedPoint::Number.new(0, format)
    
    fixt.hex = "f0"
    fixt.source.should    == -1.0
    fixt.to_f.should      == -1.0
    fixt.to_s.should      == "-1.0"
    fixt.to_i.should      == -1
    fixt.frac.should      == 0.0
    fixt.to_h.should      == "f0"
    fixt.to_b.should      == "1111.0000"
    fixt.overflow?.should  == false
    fixt.underflow?.should == false
  end

  it "hex= should ignore masb past word length" do
    format = FixedPoint::Format.new(1, 4, 4)
    fixt   = FixedPoint::Number.new(0, format)
    
    fixt.hex = "f80"
    fixt.source.should    == -8.0
    fixt.to_f.should      == -8.0
    fixt.to_s.should      == "-8.0"
    fixt.to_i.should      == -8
    fixt.frac.should      == 0.0
    fixt.to_h.should      == "80"
    fixt.to_b.should      == "1000.0000"
    fixt.overflow?.should  == false
    fixt.underflow?.should == false
  end

  it "Bin= check 0x removal " do
    format = FixedPoint::Format.new(1, 4, 4)
    fixt   = FixedPoint::Number.new(0, format)
    
    fixt.hex = "0xf0"
    fixt.source.should    == -1.0
    fixt.to_f.should      == -1.0
    fixt.to_s.should      == "-1.0"
    fixt.to_i.should      == -1
    fixt.frac.should      == 0.0
    fixt.to_h.should      == "f0"
    fixt.to_b.should      == "1111.0000"
    fixt.overflow?.should  == false
    fixt.underflow?.should == false
  end

  it "Bin= check 0x removal " do
    format = FixedPoint::Format.new(1, 4, 4)
    fixt   = FixedPoint::Number.new(0, format)
    
    fixt.hex = "0xf1"
    fixt.source.should    == -0.9375
    fixt.to_f.should      == -0.9375
    fixt.to_s.should      == "-0.9375"
    fixt.to_i.should      == 0
    fixt.frac.should      == -0.9375
    fixt.to_h.should      == "f1"
    fixt.to_b.should      == "1111.0001"
    fixt.overflow?.should  == false
    fixt.underflow?.should == false
  end
end



lib/fixed_point/format.rb
module FixedPoint

  class Format

    attr_reader :signed, :int_bits, :frac_bits
    
    # Calculated attributes
    attr_reader :width
    attr_reader :resolution, :max_value, :min_value 
    attr_reader :max_int_signed, :max_int_unsigned
    attr_reader :max_frac
    attr_reader :max_signed, :max_unsigned

    def initialize(signed, int_bits, frac_bits)
      @signed    = signed
      @int_bits  = int_bits
      @frac_bits = frac_bits

      calculate_attributes(signed, int_bits, frac_bits)
    end

    def signed?
      (@signed == 1)
    end

    #Format Should hold the maxim possible values
    #not part of number (Value)
    def calculate_attributes(signed, int_bits, frac_bits)
      @width     = int_bits + frac_bits

      #Calculate Number ranges
      @resolution       = 2**(-@frac_bits)
      @max_frac         = 1 - 2**(-@frac_bits)
      @max_int_signed   = ( 2**(@int_bits - 1) - 1)
      @max_int_unsigned = ( 2**@int_bits       - 1)
      @max_signed       = @max_int_signed   + @max_frac
      @max_unsigned     = @max_int_unsigned + @max_frac

      @min_signed       = (-2**(@int_bits-1))
      @min_unsigned     = 0

      #Set Max/Min values
      if signed?
        @max_value = @max_signed
        @min_value = @min_signed 
      else 
        @max_value = @max_unsigned
        @min_value = @min_unsigned
      end

    end

  end

end #module FixedPoint


lib/fixed_point/number.rb
module FixedPoint

  class Number
    attr_reader :source #Input value
    attr_reader :format

    ############################################
    #### Note
    ## All methods which set the @source value
    ## must also set the @quantised value.
    ## Every thing else is calculated on the fly
    ## based on those 2 values
    ############################################


    def initialize(number, input_format=Format.new(1,12,20), decimal_mark=".")
      @source       = number
      @format       = input_format
      @decimal_mark = decimal_mark

      @warnings     = false

      #Now construct values based on config data
      @quantised    = quantise_value( source )
    end

    ####################################
    ### Numerical overflow flag 
    ####################################
    def overflow?
      @overflow
    end

    ####################################
    ### Numerical underflow flag 
    ####################################
    def underflow?
      @underflow
    end

    ####################################
    ### Methods to return number formats
    ####################################

    # Alias for binary.
    def bin
      binary
    end

    # Alias for hexadecimal.
    def hex
      hexadecimal
    end

    # Alias for binary.
    def to_b
      binary
    end

    # Alias for hexadecimal.
    def to_h
      hexadecimal
    end

    #Floating Point form of the quantised value.
    def to_f
      @quantised
    end

    #Integer, integer part of quantised value.
    def to_i
      @quantised.to_i
    end

    # Fractional section of quantised value.
    def frac
      (@quantised - to_i)
    end

    def to_s
      to_f.to_s
    end

    ####################################
    ### Calculate Binary format
    ####################################
    def binary
      #Take signed quantised value and create binary string
      if (@quantised < 0) and frac.nonzero?
        # Fractional numbers not negative
        # So the integer part is 1 less than other wise would be and add 1+frac
        ret_bin_int = (@format.max_int_unsigned + to_i  )
        frac_value  = 1 + frac
      end

      if (@quantised < 0) and frac.zero?
        ret_bin_int = (@format.max_int_unsigned + to_i + 1 )
        frac_value  = frac
      end

      if @quantised >= 0
        ret_bin_int = self.to_i
        frac_value  = frac
      end

      ## Convert to binary String and extend to correct length
      ret_bin_int = ret_bin_int.to_s(2).rjust(@format.int_bits, '0')

      ## Normalise Fractional (fractional bits shifted to appear as integer)
      ret_bin_frac = Integer(frac_value * 2**@format.frac_bits)
      ret_bin_frac = ret_bin_frac.to_s(2).rjust(@format.frac_bits, '0' ) 

      #Decide if we need to add Decimal( Binary ) Point
      if @format.frac_bits > 0
        binary_string = ret_bin_int +  @decimal_mark + ret_bin_frac
      else 
        binary_string = ret_bin_int
      end

      return binary_string
    end 

    ####################################
    ### Calculate Hexadecimal format
    ####################################
    def hexadecimal
      #Clean Binary code (remove _ - . etc)
      clean_binary = to_b.scan(/[01]/).join('')

      #Convert to unsigned int then to hex
      hex          = clean_binary.to_i(2).to_s(16)
      hex_chars    = (@format.width/4.0).ceil

      ## Extend to the correct length
      ## Negative numbers will already have MSBs this if for small +VE
      return hex.rjust(hex_chars, '0')
    end

    ####################################
    ### Methods to set value from fixed point format
    ####################################
    def binary=(text)
      if text.match(/([01]*)(.?)([01]*)/ )
        set_int       = $1
        int_bits      = $1.size 

        @decimal_mark = $2
        set_frac      = $3
        frac_bits     = $3.size

        #TODO Warn if the number of bits supplied does not match @format 

        ## This should now create a new format type
        #  Do not change the Signed format as can not detect that from bit pattern
        @format = Format.new(@format.signed, int_bits, frac_bits)

        @source  = binary_to_float( set_int, set_frac)
        
        ## Set the Quantised value
        @quantised = @source

        return binary
      else 
        $stderr.puts "ERROR invalid binary input \(#{text}\)"
        return nil
      end
    end

    ## TODO this feature for version 0.2.1
    def hex=( text )
     
      ## Hex numbers do not get delimited

      ## Strip leading \d+'h (Verilog denominator)
      
      # bin= resets the wordlength
      # This should not be done for hex as the exact wordlength can not be calculated
      #if text.match(/([0-9a-fA-F]*)(.?)([0-9a-fA-F]*)/ )
        #set_int       = $1
        #int_bits      = $1.size * 4
        #@decimal_mark = $2
        #set_frac      = $3
        #frac_bits     = $3.size * 4
        #Change from hex to binary bit pattern
        #set_int  = set_int.to_i(16).to_s(2)
        #set_frac = set_frac.to_i(16).to_s(2)
      
      if text.match(/(0x)?([0-9a-fA-F]*)/ )
        binary_bits = $2.to_i(16).to_s(2)

        #Ensure min length (for reliability manipulating string)
        binary_bits = binary_bits.rjust(@format.int_bits, "0")
        if @format.frac_bits > 0
          set_frac = binary_bits[-@format.frac_bits, @format.frac_bits]
        else
          set_frac = ""
        end
        
        set_int  = binary_bits[-(@format.width)..-(@format.frac_bits+1)]

        ## Was the input word length too Long
        #oversized_bits = binary_bits.size - @format.width 
        #if oversized_bits > 0
        #  puts "Oversized Hex Value"
        #  puts "Extra bits #{oversized_bits}"
        #  puts "The extra bits binary_bits #{binary_bits[0,oversized_bits]} MSB was #{binary_bits[oversized_bits]}"
        #end

        @source    = binary_to_float( set_int, set_frac )
        @quantised = @source

      else
        $stderr.puts "ERROR invalid hex input \(#{text}\)"
        return nil
      end
    end


    #TODO
    #def log
    #def log2

    def warnings( val=true )
      @warnings = val
    end

    private

    def check_for_overflow_underflow( source, format)
      overflow  = false
      underflow = false

      #WARN +VE
      if (source > 0) and (source > format.max_value)
        puts "WARNING Maximum number is #{format.max_value} input was #{source}" if @warnings 
        overflow = true
      end

      ##WARN -VE
      if (source < 0) and (source < format.min_value)
        puts "WARNING Minimum number is #{format.min_value} input was #{source}" if @warnings
        underflow = true
      end

      return [overflow, underflow]
    end


    def quantise_value( source )
      #Overflow / Underflow flags
      @overflow, @underflow = check_for_overflow_underflow( source, @format)
      
      ## Create fractional only number
      source_frac = source - source.to_i

      #Logic for fractional negative numbers is different
      if (( source < 0) and source_frac.nonzero? )
        #Integer bits become 1 more negative
        number_int  = Integer( source )-1

        #The @fractional part inverts so int+frac = original number
        number_frac = ( source - number_int)
      else
        # Create Integer only number
        number_int  = source.to_i
        
        #Fractional Part
        number_frac = source_frac
      end


      if overflow?
        if @format.signed?
          number_int  = @format.max_int_signed
          number_frac = @format.max_frac
        else
          number_int  = @format.max_int_unsigned
          number_frac = @format.max_frac
        end
      end

      if underflow?
        if @format.signed?
          number_int  = 2**(@format.int_bits-1)
          number_frac = 0
        else
          number_int  = 0
          number_frac = 0
        end
      end

      ##Roll the integer number over the number space so binary conversion gives correct Twos complement
      if number_int < 0
        number_int = @format.max_int_unsigned + (number_int + 1)
      end

      ###################################
      ### Quantized Fractional value
      ###################################
      # Normalise, represent as integer
      # Fractional data is removed by the integer.
      # We are only left with @format.frac_bits worth of data
      number_frac_quant = (number_frac * 2**@format.frac_bits).to_i / (2**@format.frac_bits).to_f

      # TODO Comment on what this is actually doing
      # some sought of signed conversion
      if (@source < 0) and number_frac_quant.nonzero?
        number_int        = number_int - @format.max_int_unsigned
        number_frac_quant = number_frac_quant - 1
      end

      if (@source < 0) and number_frac_quant.zero?
        number_int        = number_int - @format.max_int_unsigned - 1
        number_frac_quant = 0
      end

      return (number_int + number_frac_quant)
    end ## quantise_input



    ##
    def binary_to_float(int_part, frac_part ) 
      index    = 0
      value    = 0.0
      int_part.reverse.each_char do |x|
        if x == "1"
          #If input is signed then MSB is negative
          if ((index + 1) == @format.int_bits) and (@format.signed?)
            value = value + -2**index 
          else  
            value = value + 2**index 
          end
        end
        index = index + 1
      end

      index = 1
      frac_part.each_char do |x|
        if x == "1"
          value = value + 2**-index
        end
        index = index + 1
      end

      return value
    end

   
    ## Taking methd out until covered by tests
    #def normalised
    #  #This use to be only for positive numbers
    #
    #  # This function shifts the fixedpoint number 
    #  #   so it can be represented as an integer.
    #  return  Integer((@quantised)*(2**@format.frac_bits))
    #end

    def FixedPoint_debug(msg)
      if true == false
        puts msg
      end
    end
  end #class number
end #module FixedPoint


examples/table_generation.rb

begin
  require 'rubygems'
  require 'fixed_point'
rescue
  require_relative '../lib/fixed_point'
end


puts
puts 'Unsigned 4 bit Integer'

#Unsigned, 4 bit integer
format = FixedPoint::Format.new(0,4,0)

(0...16).each do |num|
  fixt   = FixedPoint::Number.new(num, format)
  puts "#{format.width}'b#{fixt.to_b}  // #{fixt.to_f} "
end



puts
puts 'Signed 8 bit (6 bit int, 2 bit frac)'

format = FixedPoint::Format.new(1,4,4)

table_for_conversion = [ 
  2.7505, 2.5, 2.25, 2, 0.5, 0, -0.5, -4.5   
]

table_for_conversion.each do |num|
  fixt   = FixedPoint::Number.new(num, format, '_')
  puts "#{format.width}'b#{fixt.to_b}  // Quantised #{fixt.to_f}  Source #{fixt.source}"
end


puts
puts 'As above, in hex and better string formatting in comments'

## NB "%min_length.frac_length" % float_number, min_length wil lpad with saces as required
table_for_conversion.each do |num|
  fixt   = FixedPoint::Number.new(num, format, '_')
  puts "#{format.width}'h#{fixt.to_h}  // Quantised %7.4f  Source %7.4f" % [ fixt.to_f, fixt.source]
end

